src/main/scala/operations/backfill/MilvusBackfill.scala

```diff
@@ -0,0 +1,569 @@
+  ): Either[BackfillError, Map[Long, SegmentBackfillResult]] = {
+
+    try {
+      // Prepare data: select only needed columns and add segment_id for partitioning
+      val preparedDF = joinedDF
+        .select((Seq("segment_id", "row_offset") ++ newFieldNames).map(col): _*)
+
+      // Repartition by segment_id to ensure all rows of same segment go to same partition
+      // Sort within partitions to maintain row_offset order
+      val repartitionedDF = preparedDF
+        .repartition(col("segment_id"))
+        .sortWithinPartitions("segment_id", "row_offset")
+
+      // Get the schema for new fields only (without segment_id and row_offset)
+      val targetSchema = org.apache.spark.sql.types.StructType(
+        newFieldNames.map(fieldName =>
+          preparedDF.schema.fields.find(_.name == fieldName).get
+        )
+      )
+
+      // Broadcast configuration to executors
+      val broadcastConfig = spark.sparkContext.broadcast(config)
+      val broadcastCollectionID = spark.sparkContext.broadcast(collectionID)
+      val broadcastPartitionID = spark.sparkContext.broadcast(partitionID)
+      val broadcastTargetSchema = spark.sparkContext.broadcast(targetSchema)
+
+      // Get the underlying RDD[InternalRow] for efficient processing
+      val internalRowRDD = repartitionedDF.queryExecution.toRdd
+
+      // Process each partition (which may contain one or more segments)
+      // Each segment will be written by exactly one FFI writer
+      val results = internalRowRDD.mapPartitions { iter =>
+        if (!iter.hasNext) {
+          // Empty partition
+          Iterator.empty
+        } else {
+          val cfg = broadcastConfig.value
+          val collID = broadcastCollectionID.value
+          val partID = broadcastPartitionID.value
+          val schema = broadcastTargetSchema.value
+
+          // Group rows by segment_id within this partition
+          val segmentGroups = groupRowsBySegmentId(iter)
+
+          // Process each segment in this partition
+          segmentGroups.map { case (segmentID, rows) =>
+            processSegmentWithWriter(
+              segmentID = segmentID,
+              rows = rows,
+              collectionID = collID,
+              partitionID = partID,
+              config = cfg,
+              targetSchema = schema
+            )
+          }
+        }
+      }.collect()
+
+      // Cleanup broadcast variables
+      broadcastConfig.unpersist()
+      broadcastCollectionID.unpersist()
+      broadcastPartitionID.unpersist()
+      broadcastTargetSchema.unpersist()
+
+      // Check for failures
+      val failures = results.filter(_._2.isDefined)
+      if (failures.nonEmpty) {
+        val firstFailure = failures.head
+        val error = firstFailure._2.get
+        return Left(WriteError(
+          segmentId = firstFailure._1.segmentId,
+          outputPath = firstFailure._1.outputPath,
+          message = s"Failed to write ${failures.length} segment(s): ${error.getMessage}",
```

Comment:
Do we need to handle data clean up? I guess not, just utilize MIlvus' orphan clean mechanism